File: src/main/java/com/github/alexandrecarlton/idea/settings/layout/configurations/docker/DockerComposeConfigurationSettings.java
package com.github.alexandrecarlton.idea.settings.layout.configurations.docker;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.github.alexandrecarlton.idea.settings.layout.configurations.ConfigurationSettings;
import java.nio.file.Path;
import java.util.List;
import java.util.Optional;
import org.immutables.value.Value;

@Value.Immutable
@JsonDeserialize(as = ImmutableDockerComposeConfigurationSettings.class)
public interface DockerComposeConfigurationSettings extends ConfigurationSettings {

  Optional<List<Path>> composeFiles();

  Optional<List<String>> services();

  Optional<List<DockerEnvironmentVariable>> environmentVariables();

  Optional<DockerComposeConfigurationOptionsSettings> options();
}

File: src/main/kotlin/com/github/alexandrecarlton/idea/settings/layout/configurations/docker/DockerComposeConfigurationSettings.kt
package com.github.alexandrecarlton.idea.settings.layout.configurations.docker

import com.fasterxml.jackson.annotation.JsonProperty
import com.github.alexandrecarlton.idea.settings.layout.configurations.ConfigurationSettings
import com.github.alexandrecarlton.idea.settings.layout.configurations.common.before_launch.BeforeLaunchConfigurationSettings
import java.nio.file.Path

data class DockerComposeConfigurationSettings(

    @JsonProperty("Name")
    override val name: String,

    @JsonProperty("Share through VCS")
    override val shareThroughVcs: Boolean? = null,

    @JsonProperty("Before launch")
    override val beforeLaunch: List<BeforeLaunchConfigurationSettings>? = null,

    @JsonProperty("Compose file(s)")
    val composeFiles: List<Path>? = null,

    @JsonProperty("Service(s)")
    val services: List<String>? = null,

    @JsonProperty("Environment variables")
    val environmentVariables: List<DockerEnvironmentVariable>? = null,

    @JsonProperty("Options")
    val options: DockerComposeConfigurationOptionsSettings? = null
) : ConfigurationSettings
